aten/src/ATen/autocast_mode.cpp

```diff
@@ -157,28 +203,57 @@ struct WrapFunction_<CastPolicy::fp32_set_opt_dtype, Redispatch, F, Ret, guts::t
   }
 };
 
-// CastPolicy::fp32_append_dtype
+// DeviceType::CUDA CastPolicy::fp32_append_dtype
 template<class Redispatch, Redispatch* F, class Ret, class... Args>
-struct WrapFunction_<CastPolicy::fp32_append_dtype, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+struct WrapFunction_<DeviceType::CUDA, CastPolicy::fp32_append_dtype, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
   static Ret call(Args... args) {
     c10::impl::ExcludeDispatchKeyGuard no_autocast(DispatchKey::Autocast);
     at::ScalarType out_type = type_from_firstarg(at::kFloat, args...);
     return (*F)(args..., out_type);
   }
 };
 
-// CastPolicy::promote
+// DeviceType::CUDA CastPolicy::promote
 template<class Redispatch, Redispatch* F, class Ret, class... Args>
-struct WrapFunction_<CastPolicy::promote, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+struct WrapFunction_<DeviceType::CUDA, CastPolicy::promote, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
   static Ret call(Args... args) {
     c10::impl::ExcludeDispatchKeyGuard no_autocast(DispatchKey::Autocast);
     auto to_type = promote_type(at::kHalf, args...);
     return (*F)(cached_cast(to_type, args)...);
   }
 };
 
+// DeviceType::CPU CastPolicy::user_defined_dtype
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<DeviceType::CPU, CastPolicy::user_defined_dtype, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocast(DispatchKey::AutocastCPU);
+    return (*F)(cached_cast(DeviceType::CPU, autocast_cpu_dtype, args)...);
```

Comment:
You can totally deduplicate this code. Define a helper function to translate DeviceType into DispatchKey (CPU to AutocastCPU), rewrite `cached_cast` to directly query `autocast_cpu_dtype` rather than taking it as argument, then you can polymorphically pass device type on to cached_cast and reuse implementations.

Reply:
Oops I should have put this comment on the fp32 struct below, since technically this particular specialization isn't shared. But see above for more...

Reply:
Did you mean to remove the `to_type` argument in `cached cast`? Then how can we judge whether the case is cased from fp32 to bf16 or cast back from bf16 to fp32?

Reply:
You'll base it off of the DeviceType.

Reply:
yes, for the cast policy shared by different device we can define the WrapFunction_ template as 
```
template<DeviceType device_type, class Redispatch, Redispatch* F, class Ret, class... Args>
struct WrapFunction_<CastPolicy::fp32, device_type, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {}
```
Then it can be shared by different devices to save the duplication code.